Feature/ae 1979 sakkopaatos valitusajan odotus ja umpeutuminen toimenpide

solita-juhohaa opened the pull request.

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/valvonta_kaytto/toimenpide.clj:
```diff
@@ -52,26 +53,29 @@
                         :penalty-decision-notice-first-mailing
-                        :penalty-decision-notice-second-mailing}))
+                        :penalty-decision-notice-second-mailing
+                        :penalty-decision-waiting-for-deadline}))
```
Ei liity tähän tickettiin, lisättiin puuttuva toimenpide.

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/valvonta_kaytto/toimenpide.clj:
```diff
@@ -52,26 +53,29 @@
                         :decision-order-hearing-letter
                         :decision-order-actual-decision
-                        :decision-order-notice-first-mailing11
+                        :decision-order-notice-first-mailing
```
Ei liity tähän tickettiin, korjattiin typo.

solita-juhohaa commented on etp-backend/src/main/sql/solita/etp/db/valvonta-kaytto.sql:
```diff
@@ -333,6 +333,11 @@ SELECT id
+SELECT id
+FROM vk_toimenpidetype
+WHERE allow_comments is true;
+
```
Käytin tätä aluksi testissä, mutta nyt sitä ei enää käytetä, kun kommenttien salliminen olikin testattu jo `solita.etp.service.valvonta-kaytto-test/find-toimenpidetyypit-test`. Jätin nyt näköjään kuitenkin, mutta kuuluisiko poistaa, kun ei käytetä?

Juholei replied:
Poista vaan

Juholei replied:
Kommenttina tähän myös, että jos ymmärsin oikein miten olisit käyttänyt tätä testissä, niin sehän olisi siinä tapauksessa tarkoittanut sitä, että tietokannassa olevaa dataa verrataan testissä tietokannassa olevaan dataan. Eli ei oikeastaan testata mitään? solita.etp.service.valvonta-kaytto-test/find-toimenpidetyypit-test testissä kun haluttu data on kirjoitettu siihen testiin niin sillä varmistetaan se, että tietokantatauluun on päätynyt se data mitä siellä halutaan olevan.

solita-juhohaa replied:
https://github.com/solita/etp-core/pull/1030/commits/67647309517d8c8faeaee0fa38b497e4d08e4f71

Tuolleen sitä käytin. Tuohon suuntaan meinasin alkaa laajentamaan tuota funktiota, mutta tajusin, että noita testataan jo..

Juholei commented on etp-backend/src/test/clj/solita/etp/valvonta_kaytto/sakkopaatos_odotetaan_valitusajan_umpeutumista_test.clj:
```diff
@@ -0,0 +1,57 @@
+    [solita.etp.test-system :as ts])
+  (:import (java.time LocalDate)))
+
+(t/use-fixtures :each ts/fixture)
```
Laita tyhjä rivi näiden väliin

Juholei commented on etp-backend/src/main/clj/solita/etp/service/valvonta_kaytto/toimenpide.clj:
```diff
@@ -84,9 +88,17 @@
\ No newline at end of file
+  (boolean (:recipient-answered osapuoli-specific-data)))
+
+(defn allow-comments? [db type-id]
```
Tätäkään ei nyt ilmeisesti käytetä kuten ei tuota alempana olevaa tietokantaqueryä?

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/valvonta_kaytto/toimenpide.clj:
```diff
@@ -23,8 +23,12 @@
    11 :decision-order-notice-bailiff
    12 :decision-order-waiting-for-deadline
    14 :penalty-decision-hearing-letter
+   15 :change-when-implement-penalty-decision-actual-decision
```
Täytyi lisätä kaikille kannassa oleville toimenpiteille jokin placeholder tähän. Näitä on nyt siis idt 15, 18, 21. Laitoin eteen **change-when-implement**.

Id 19 on tämän ticketin toimenpide.

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/valvonta_kaytto/toimenpide.clj:
```diff
@@ -52,26 +56,29 @@
                         :decision-order-notice-second-mailing
-                        :decision-order-notice-bailiff}))
+                        :decision-order-notice-bailiff
+                        :decision-order-waiting-for-deadline}))
```
Ei liity tähän tickettiin, lisättiin puuttuva toimenpide.

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
+      (t/is (true? (toimenpide/asha-toimenpide? {:type-id 14})))))
+
+  (t/testing "Sakkopäätös / varsinainen päätös"
+    (t/testing "has type-id 15 and type-key :change-when-implement-penalty-decision-actual-decision"
```
Lisätty testi, vaikka toteuttamatta vielä.

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
 
+  (t/testing "Sakkoluettelon lähetys menossa"
+    (t/testing "has type-id 21 and type-key :sakkoluettelo-delivery-ongoing"
+      (t/is (= (toimenpide/type-key 21) :change-when-implement-sakkoluettelo-delivery-ongoing)))
```
Lisätty testi, vaikka toteuttamatta vielä.

solita-juhohaa committed:
AE-1979: Allow comments for Sakkopäätös / valitusajan odotus ja umpeutuminen

solita-juhohaa committed:
AE-1979: Add type-key for penalty-decision-waiting-for-deadline and placeholders for missing

solita-juhohaa committed:
AE-1979: Add penalty decision-waiting-for-deadline to predicates and fix problems

Fixed typo in asha-toimenpide? and added missing toimenpide from sakkopaatos-toimenpide?

solita-juhohaa committed:
AE-1979: Add api test for sakkopaatos-odotetaan-valitusajan-umpeutumista

solita-juhohaa committed:
AE-1979: Test toimenpiteet

Test that:
- toimenpide in database has been assigned a type-key
- there are no duplicate type-keys
- predicates return what's expected for each toimenpide
- type-key and type-id are what's expected for toimenpide

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
 
+  (t/testing "Sakkoluettelon lähetys menossa"
+    (t/testing "has type-id 21 and type-key :change-when-implement-sakkoluettelo-delivery-ongoing"
+      (t/is (= (toimenpide/type-key 21) :change-when-implement-sakkoluettelo-delivery-ongoing)))
```
Lisätty testi, vaikka toteuttamatta vielä.

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
 
+  (t/testing "Sakkopäätös / tiedoksianto (Haastemies)"
+    (t/testing "has type-id 18 and type-key :change-when-implement-penalty-decision-notice-bailiff"
+      (t/is (= (toimenpide/type-key 18) :change-when-implement-penalty-decision-notice-bailiff)))
```
Lisätty testi, vaikka toteuttamatta vielä.

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
+
+  (t/testing "Sakkopäätös / varsinainen päätös"
+    (t/testing "has type-id 15 and type-key :change-when-implement-penalty-decision-actual-decision"
+      (t/is (= (toimenpide/type-key 15) :change-when-implement-penalty-decision-actual-decision)))
```
Lisätty testi, vaikka toteuttamatta vielä.

Juholei commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
+                                    (filter type-key-exists-and-is-keyword?)
+                                    (map toimenpide/type-key))]
+      (t/testing "type-key keywords exist for all ids in the database"
+        (doseq [test-case (map #(t/testing (str "Toimenpide with id " %1 " in database has a type-key keyword")
```
En ihan hoksaa tämän mapin tarpeellisutta tässä.

solita-juhohaa replied:
Mapataan siis funktio
```clojure
#(t/testing (str "Toimenpide with id " %1 " in database has a type-key keyword")
  (t/is (true? (type-key-exists-and-is-keyword? %1))))
```
tuohon
```clojure
type-ids-from-database
```
jolloin saadaan test-case jokaiselle kannassa olevalle idlle.

Täytyy myös käyttää `doseq`, koska muuten noita ei evaluoida.

Kokeilin, että onko tässä jokin juju ja kirjoitin saman
```clojure
      (doall (for [id type-ids-from-database
                   :let [test-case (#(t/testing (str "Toimenpide with id " %1 " in database has a type-key keyword")
                                       (t/is (true? (type-key-exists-and-is-keyword? %1)))) id)]]
               test-case))
```
joka taitaa olla vain sama koodi kirjoitettuna `for`-macrolla...

En nyt tiedä onko tuo funktio selvästi/oikein kirjoitettu. Olisiko tähän jokin selvempi tapa kirjoittaa tai käsittelenkö noita lazy sequencejä järkevästi? 🤔

Juholei replied:
Onnistuisiko tuossa doseqissä käydä läpi vain id:eitä itse testing ja sen sisällä is olisi sen doseqin sisällä?

Juholei replied:
Jotenkin siis
`(doseq [id type-ids-from-database]
  (t/testing (str "Toimenpide with id " id " in database has a type-key keyword")
    (t/is (true? (type-key-exists-and-is-keyword? id)))))`

solita-juhohaa replied:
Joo, tein noin.

Juholei commented on etp-backend/src/test/clj/solita/etp/service/valvonta_kaytto/toimenpide_test.clj:
```diff
@@ -61,13 +62,214 @@
+(t/deftest all-toimenpiteet-test
+  (t/testing "All toimenpiteet"
+    (let [type-ids-from-database (map :id (valvonta-kaytto/find-toimenpidetyypit ts/*db*))
+          type-key-exists-and-is-keyword? #(try (keyword? (toimenpide/type-key %1)) (catch Exception _ false))
```
type-key-funktio heittää poikkeuksen jos avainta ei löydy. Pitääkö tässä catchata poikkeus jostain syystä? Tuntuisi oikeammalta antaa tässä poikkeuksen lentää jos avainta ei löydy, jolloin testi feilaisi sen takia.

solita-juhohaa replied:
Muistaakseni se silloin erroraa ja lopettaa suorittamisen, jolloin ei saanut selville montako type-keytä puuttuu, mutta testaanpa vielä..

solita-juhohaa replied:
Ei pidä paikkaansa tuo äskeinen väitteeni. Otan tuon try-catchin pois..
